java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -0,0 +1,157 @@
+      .filter(entry -> scannedPackages.stream().noneMatch(entry.getKey()::contains))
+      .forEach(entry -> entry.getValue().forEach(reporter::reportIssue));
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic() || !(context instanceof AnalyzerMessageReporter)) {
+      return;
+    }
+
+    ClassTree classTree = (ClassTree) tree;
+    String classPackageName = packageNameOf(classTree);
+
+    Optional<AnnotationTree> componentScanAnnotation = classTree.modifiers().annotations().stream()
+      .filter(SpringBeansShouldBeAccessibleCheck::isComponentScan).findFirst();
+    if (componentScanAnnotation.isPresent()) {
+      componentScanAnnotation.get().arguments().forEach(this::addToScannedPackages);
+    } else if (hasAnnotation(classTree, SPRING_BOOT_APP_ANNOTATION)) {
+      scannedPackages.add(classPackageName);
+    } else if (hasAnnotation(classTree, SPRING_BEAN_ANNOTATIONS)) {
+      addMessageToMap(classPackageName, classTree);
+    }
+  }
+
+  private void addMessageToMap(String classPackageName, ClassTree classTree) {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    IdentifierTree className = classTree.simpleName();
+    if (className != null) {
+      AnalyzerMessage analyzerMessage = reporter.createAnalyzerMessage(this, className, String.format(MESSAGE_FORMAT, className));
+      messagesPerPackage.computeIfAbsent(classPackageName, k -> new ArrayList<>());
+      messagesPerPackage.get(classPackageName).add(analyzerMessage);
+    }
+  }
+
+  private void addToScannedPackages(ExpressionTree annotationArgument) {
+    if (annotationArgument.is(Tree.Kind.ASSIGNMENT)) {
+      AssignmentExpressionTree argumentAssignment = (AssignmentExpressionTree) annotationArgument;
+      ExpressionTree argumentAssignmentVar = argumentAssignment.variable();
+      if (argumentAssignmentVar.is(Tree.Kind.IDENTIFIER) && COMPONENT_SCAN_ARGUMENTS.contains(((IdentifierTree) argumentAssignmentVar).name())) {
+        addLiteralsToScannedPackages(argumentAssignment.expression());
+      }
+    } else {
+      addLiteralsToScannedPackages(annotationArgument);
+    }
+  }
+
+  private void addLiteralsToScannedPackages(ExpressionTree packageNames) {
+    if (packageNames.is(Tree.Kind.STRING_LITERAL)) {
+      String name = ConstantUtils.resolveAsStringConstant(packageNames);
+      scannedPackages.add(name);
+    } else if (packageNames.is(Tree.Kind.NEW_ARRAY)) {
+      for (ExpressionTree p : ((NewArrayTree) packageNames).initializers()) {
+        String name = ConstantUtils.resolveAsStringConstant(p);
+        scannedPackages.add(name);
+      }
+    }
+  }
+
+  private static String packageNameOf(ClassTree classTree) {
+    Tree classTreeParent = classTree.parent();
+    if (classTreeParent != null && classTreeParent.is(Tree.Kind.COMPILATION_UNIT)) {
+      return PackageUtils.packageName(((CompilationUnitTree) classTreeParent).packageDeclaration(), ".");
+    }
+    return "";
```

Comment:
So we implicitly ignore inner classes which would be beans? Why? (Not sure it's possible, from a spring perspective). Worth adding a comment if it's not possible, or handle the inner classes.